src/cryptsetup/cryptsetup-fido2.c

```diff
@@ -121,42 +135,42 @@ int acquire_fido2_key(
         }
 }
 
-int find_fido2_auto_data(
+int acquire_fido2_key_auto(
                 struct crypt_device *cd,
-                char **ret_rp_id,
-                void **ret_salt,
-                size_t *ret_salt_size,
-                void **ret_cid,
-                size_t *ret_cid_size,
-                int *ret_keyslot,
-                Fido2EnrollFlags *ret_required) {
-
-        _cleanup_free_ void *cid = NULL, *salt = NULL;
-        size_t cid_size = 0, salt_size = 0;
-        _cleanup_free_ char *rp = NULL;
-        int r, keyslot = -1;
+                const char *name,
+                const char *friendly_name,
+                const char *fido2_device,
+                const char *key_file,
+                size_t key_file_size,
+                uint64_t key_file_offset,
+                usec_t until,
+                bool headless,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                AskPasswordFlags ask_password_flags) {
+
+        _cleanup_free_ void *cid = NULL;
+        size_t cid_size = 0;
+        int r = -ENOENT;
         Fido2EnrollFlags required = 0;
 
         assert(cd);
-        assert(ret_salt);
-        assert(ret_salt_size);
-        assert(ret_cid);
-        assert(ret_cid_size);
-        assert(ret_keyslot);
-        assert(ret_required);
 
         /* Loads FIDO2 metadata from LUKS2 JSON token headers. */
 
         for (int token = 0; token < sym_crypt_token_max(CRYPT_LUKS2); token ++) {
                 _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
                 JsonVariant *w;
+                _cleanup_free_ void *salt = NULL;
+                _cleanup_free_ char *rp = NULL;
+                size_t salt_size = 0;
                 int ks;
 
-                r = cryptsetup_get_token_as_json(cd, token, "systemd-fido2", &v);
-                if (IN_SET(r, -ENOENT, -EINVAL, -EMEDIUMTYPE))
+                int res_json = cryptsetup_get_token_as_json(cd, token, "systemd-fido2", &v);
```

Comment:
why is a new variable used instead of `r`?

Reply:
Because we want to preserve the error codes of previous calls to `acquire_fido2_key` even if later calls to `cryptsetup_get_token_as_json` fail. This loop will only terminate if `acquire_fido2_key` succeeds or we exhaust all LUKS key slots, in which case `cryptsetup_get_token_as_json` fails with one of the error codes checked below. When that happens, we still want this function to return the error code of the last actual call to `acquire_fido2_key`.

Reply:
please do this the other way round, i.e. use "r" for the local errors, and collect the final error in some variable "ret" or "rr" or so. That'd be more in line how we usually do things.

i.e. `r` should always be good to use for every use, and if you want to keep the error for longer, than that's the special case that should get its own variable.